clippy_lints/src/use_self.rs

```diff
@@ -467,3 +467,18 @@ fn should_lint_ty(hir_ty: &hir::Ty<'_>, ty: Ty<'_>, self_ty: Ty<'_>) -> bool {
         }
     }
 }
+
+fn in_macro_recursively(cx: &LateContext<'_>, hir_id: HirId) -> bool {
+    let map = cx.tcx.hir();
+    if map.opt_span(hir_id).map_or(false, in_macro) {
+        return true;
+    }
+
+    for (parent_id, _) in map.parent_iter(hir_id) {
+        if map.opt_span(parent_id).map_or(false, in_macro) {
+            return true;
+        }
+    }
+
+    false
+}
```

Comment:
This looks expensive and doesn't seem like the right solution to me. We probably need to check `in_macro` somewhere while descending into the HIR rather than the other way around.

Reply:
I agree, we shouldn't have to traverse the HIR to determine if something comes from a macro expansion. Is this maybe a bug in rustc, not tracking the origin of this span correctly?